### PR TITLE
Correctly pass through reference modifiers when extracting a variable

### PR DESCRIPTION
**Pull request by Vannevelj**
Fixes https://github.com/rust-analyzer/rust-analyzer/issues/10034

This will parse the field expression and look at whether it is marked `&` or `&mut` and include a modifier if appropriate. The original issue only mentions `&mut params` but I've found that this issue also occurs for `&mut locals` as well as `&params` and `&locals` so I've also added tests for them.

I'd definitely be interested in hearing where I can make my code more idiomatic for Rust.

### CODE REVIEW

File: crates/ide_assists/src/handlers/extract_variable.rs [hunk 1]
```diff
@@ -146,6 +166,29 @@ fn valid_target_expr(node: SyntaxNode) -> Option<ast::Expr> {
...
+    Some(ctx.sema.type_of_expr(&receiver)?.original())
+}
+
+fn get_receiver(expression: ast::Expr) -> Option<ast::Expr> {
+    match expression {
+        ast::Expr::FieldExpr(field) if field.expr().is_some() => {
+            let nested_expression = &field.expr()?;
+            get_receiver(nested_expression.to_owned())
+        }
+        ast::Expr::PathExpr(_) => Some(expression),
```
Comment (Vannevelj) [inline crates/ide_assists/src/handlers/extract_variable.rs:180]: Is it correct to assume that in the expression `aaaa.bbbb.cccc.xxxx()`, (of type `FieldExpr`), if I keep going deeper into the nested receivers eventually I will end up at `aaaa` which is of type `PathExpr`? 

That's the type I seemed to get so I used that as the end condition of the recursion.

File: crates/ide_assists/src/handlers/extract_variable.rs [hunk 2]
```diff
@@ -52,6 +52,18 @@ pub(crate) fn extract_variable(acc: &mut Assists, ctx: &AssistContext) -> Option
...
+        if receiver_type.is_mutable_reference() {
+            RefKind::MutRef
+        } else if receiver_type.is_reference() {
+            RefKind::Ref
+        } else {
+            RefKind::None
+        }
+    } else {
+        RefKind::None
+    };
```
Comment (Veykril) [inline crates/ide_assists/src/handlers/extract_variable.rs:65]:
```suggestion
    let ref_kind = match get_receiver_type(&ctx, &to_extract) {
        Some(receiver_type) if receiver_type.is_mutable_reference() => RefKind::MutRef,
        Some(receiver_type) if receiver_type.is_reference() => RefKind::Ref,
        _ => RefKind::None,
    };
```

File: crates/ide_assists/src/handlers/extract_variable.rs (hunk 1 above)
Comment (Veykril) [reply crates/ide_assists/src/handlers/extract_variable.rs:181]: in the general case yes, but this is not required, so this match should only have to branches, the one specializing on `ast::Expr::FieldExpr` and the other as the default just returning the current expression.

File: crates/ide_assists/src/handlers/extract_variable.rs [hunk 3]
```diff
@@ -146,6 +166,30 @@ fn valid_target_expr(node: SyntaxNode) -> Option<ast::Expr> {
     }
 }
 
+fn get_receiver_type(ctx: &AssistContext, expression: &ast::Expr) -> Option<hir::Type> {
+    let receiver = get_receiver(expression.to_owned())?;
```
Comment (Veykril) [inline crates/ide_assists/src/handlers/extract_variable.rs:170]:
```suggestion
    let receiver = get_receiver(expression.clone())?;
```
Using `clone` is more idiomatic(at least for our codebase) than using `to_owned` to go from `&T` to `T`

File: crates/ide_assists/src/handlers/extract_variable.rs (hunk 1 above)
Comment (Vannevelj) [reply crates/ide_assists/src/handlers/extract_variable.rs]: Thanks! I've made that change as well as the others you remarked

**Comment by Veykril** [comment]
Thanks!
bors r+

**Pull request merged**